python/ray/data/expressions.py

```diff
@@ -0,0 +1,181 @@
+
+    def __lt__(self, other):
+        return self._bin(other, "lt")
+
+    def __ge__(self, other):
+        return self._bin(other, "ge")
+
+    def __le__(self, other):
+        return self._bin(other, "le")
+
+    def __eq__(self, other):
+        return self._bin(other, "eq")
+
+    # boolean
+    def __and__(self, other):
+        return self._bin(other, "and")
+
+    def __or__(self, other):
+        return self._bin(other, "or")
+
+    # Rename the output column
+    def alias(self, name: str) -> "AliasExpr":
+        return AliasExpr(self, name)
+
+
+@DeveloperAPI
+@dataclass(frozen=True, eq=False)
+class ColumnExpr(Expr):
+    name: str
+
+
+@DeveloperAPI
+@dataclass(frozen=True, eq=False)
+class LiteralExpr(Expr):
+    value: Any
+
+
+@DeveloperAPI
+@dataclass(frozen=True, eq=False)
+class BinaryExpr(Expr):
+    op: str
+    left: Expr
+    right: Expr
+
+
+@DeveloperAPI
+@dataclass(frozen=True, eq=False)
+class AliasExpr(Expr):
+    expr: Expr
+    name: str
+
+
+# ──────────────────────────────────────
+#  User helpers
+# ──────────────────────────────────────
+
+
+@DeveloperAPI
+def col(name: str) -> ColumnExpr:
+    """Reference an existing column."""
+    return ColumnExpr(name)
+
+
+@DeveloperAPI
+def lit(value: Any) -> LiteralExpr:
+    """Create a scalar literal expression (e.g. lit(1))."""
+    return LiteralExpr(value)
+
+
+# ──────────────────────────────────────
+#  Local evaluator (pandas batches)
+# ──────────────────────────────────────
+# This is used by Dataset.with_columns – kept here so it can be re-used by
+# future optimised executors.
+_PANDAS_OPS: Dict[str, Callable[[Any, Any], Any]] = {
+    "add": operator.add,
+    "sub": operator.sub,
+    "mul": operator.mul,
+    "div": operator.truediv,
+    "gt": operator.gt,
+    "lt": operator.lt,
+    "ge": operator.ge,
+    "le": operator.le,
+    "eq": operator.eq,
+    "and": operator.and_,
+    "or": operator.or_,
+}
+
+_NUMPY_OPS: Dict[str, Callable[[Any, Any], Any]] = {
+    "add": np.add,
+    "sub": np.subtract,
+    "mul": np.multiply,
+    "div": np.divide,
+    "gt": np.greater,
+    "lt": np.less,
+    "ge": np.greater_equal,
+    "le": np.less_equal,
+    "eq": np.equal,
+    "and": np.logical_and,
+    "or": np.logical_or,
+}
+
+_ARROW_OPS: Dict[str, Callable[[Any, Any], Any]] = {
+    "add": pc.add,
+    "sub": pc.subtract,
+    "mul": pc.multiply,
+    "div": pc.divide,
+    "gt": pc.greater,
+    "lt": pc.less,
+    "ge": pc.greater_equal,
+    "le": pc.less_equal,
+    "eq": pc.equal,
+    "and": pc.and_,
+    "or": pc.or_,
+}
+
+
+def _eval_expr_recursive(expr: Expr, batch, ops: Dict[str, Callable]) -> Any:
+    """Generic recursive expression evaluator."""
+    if isinstance(expr, ColumnExpr):
+        return batch[expr.name]
+    if isinstance(expr, LiteralExpr):
+        return expr.value
+    if isinstance(expr, BinaryExpr):
+        return ops[expr.op](
+            _eval_expr_recursive(expr.left, batch, ops),
```

Comment:
We'd need to separate out unresolved exprs (arbitrary AST with unresolved refs) and resolved exprs (bound to a schema)

Standalone resolution process would allow in particular surface errors in a more manageable way.

We can do that split in a follow-up PR (adding ResolvedExpr), just leave a TODO here